source/modulo_core/src/Component.cpp

```diff
@@ -28,20 +28,23 @@ void Component::add_predicate(const std::shared_ptr<state_representation::Predic
   }
   // add the predicate to the map
   this->predicates_.insert(std::make_pair(predicate_name, predicate));
+  // add the publisher predicate
+  this->add_publisher<std_msgs::msg::Bool>("/predicates/" + std::string(this->get_name()) + "/" + predicate_name, predicate, true);
 }
 
 void Component::add_predicate(const std::string& predicate_name, const std::function<bool(void)>& predicate_function) {
   // crate a predicate with named prefixed with the action name
   auto predicate = std::make_shared<state_representation::Predicate>(predicate_name);
   this->add_predicate(predicate);
   // insert the predicate function in the list
-  this->predicate_functions_.push_back(std::make_pair(predicate_name, predicate_function));
+  this->predicate_functions_.insert(std::make_pair(predicate_name, predicate_function));
```

Comment:
I'm tempted to say that `insert_or_assign` would be better here, so that we can overwrite the callback function for an existing predicate if needed

Reply:
Yes this seems fair. It could actually be the default behavior for all handlers in `Cell`

Reply:
Ah well I am not sure actually. So the problem is we now directly assign the publisher. However, ros will throw an error if you assign a new publisher / subscriber on the same topic. So I actually think the proper behavior would be to throw an exception

Reply:
I actually do that line 26 in the `add_predicate` function called the line above (38):

```cpp
if (this->predicates_.find(predicate_name) != this->predicates_.end()) {
    throw exceptions::PredicateAlreadyRegisteredException(predicate_name);
  }
```

Reply:
I don't want to overwrite the handler (since removing and recreating the handler on the same interface / topic is not useful), rather just the callback function in this line

Reply:
I understand that. But this is not possible withing this call. If you do:

```cpp
add_predicate("pred", callbackA)
add_predicate("pred", callbackB)
```

It will throw a `PredicateAlreadyRegisteredException` coming from the `add_predicate(predicate)` call line 38.

Reply:
This is resolved now by catching the exception and changing the call order